Lesson18(final)

JuicyFlash opened the pull request:
Реализовано задание бейджи

JuicyFlash committed:
Реализована модель Badges
Реализована модель GivenBadges (выданные баджи)

JuicyFlash committed:
Реализован сервис выдачи баджей

JuicyFlash committed:
Сервис выдачи баджей встроен в приложение
Добавлен список баджей для пользователя
Добавлен список полученных баджей пользователем
Реализованы основные действия с баджами - удаление, реактирование, добавление (для администратора)
Локализация
Некоторые правки интерфейса приложения

JuicyFlash committed:
Правка GivenBadgesController

JuicyFlash committed:
Изменение правила (поиск по имени категории)

JuicyFlash committed:
Правка правила все тесты категории

JuicyFlash committed:
Правка правила все тесты категории. ещё

JuicyFlash committed:
Правка правила все тесты категории.

JuicyFlash committed:
Правка правила все тесты категории.

VidgarVii commented on app/controllers/admin/badges_controller.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def edit
+
+  end
```
`def edit; end`

VidgarVii commented on app/controllers/admin/badges_controller.rb:
```diff
@@ -0,0 +1,49 @@
+    @badge = Badge.find(params[:id])
+  end
+
+end
```
в конце файла надо отставлять 1 пустую строчку. Это указатель на конец файла. Сейчас система думает, что происходит обрыв файла. Даже гитхаб ругается. No new lina at end of file. Многие компиляторы тут выдадут ошибку. Руби частично компилируемый язык. И такой разрыв легко проглотит. Но в реальных проектах это как красная тряпка для быка. Так что тренируем дисциплину и больше такого не совершаем! В руби за чистоту кода!

VidgarVii commented on app/controllers/given_badges_controller.rb:
```diff
@@ -0,0 +1,7 @@
+    @badges = current_user.badges
+  end
+
+end
```
код должен быть единообразен. Используй рубокоп. К этому уроку таких стилистических ошибок уже не должно быть
```
class GivenBadgesController < ApplicationController
  def index
    @badges = current_user.badges
  end
end
```

VidgarVii commented on app/controllers/processed_tests_controller.rb:
```diff
@@ -7,7 +7,7 @@ def show
 
   def result
-
+    @badges = BadgeService.new(@processed_test).call
```
в данном решении можно постоянно обновлять данную страницу результата и получать баджи. Либо нагружать систему. 
Бадж следует выдавать после завершения теста. А он завершается не тут

VidgarVii commented on app/helpers/badges_helper.rb:
```diff
@@ -0,0 +1,11 @@
+module BadgesHelper
+  def badge_header(badge)
+    if badge.persisted?
+      #  "Edit #{test.title} Test"
```
отладочный код надо удалять
Если это документация, то её пишут выше метода

VidgarVii commented on app/helpers/badges_helper.rb:
```diff
@@ -0,0 +1,11 @@
+      I18n.t('.new_badge_header', title: badge.title)
+    end
+
+    end
```
end уехал веред

VidgarVii commented on app/javascript/utilities/badges/rule_description.js:
```diff
@@ -0,0 +1,12 @@
+    }
+})
+function fillRuleDescription(){
+    var rule_type =  document.querySelectorAll('#badge_rule_type')[0]
```
Время глобальных var-ов прошло. при объявлении переменных используй let/const. Так безопасней и код становиться понятней. Обязательно по гугли отличие!

VidgarVii commented on app/models/badge.rb:
```diff
@@ -0,0 +1,10 @@
+  validates :rule_type, presence: true
+
+
+
```
лишние отступы. Будь внимательней! Вроде бы меточь, а на работу с таким кодом уже не возьмут(. Там за каждый пробел будут ругать

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+  RULES = {finish_tests_by_category:FinishTestsByCategoryRuleSpecification,
+           finish_tests_by_level:FinishTestsByLevelRuleSpecification,
+           finish_any_test_by_first_try:FinishAnyTestByFirstTryRuleSpecification}.freeze
+  def initialize(processed_test)
```
перед константой и методом надо оставлять пустую строку

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+class BadgeService
+  RULES = {finish_tests_by_category:FinishTestsByCategoryRuleSpecification,
+           finish_tests_by_level:FinishTestsByLevelRuleSpecification,
```
неоченьприятночитатьтакойкод

VidgarVii commented on app/views/admin/badges/_form.html.erb:
```diff
@@ -0,0 +1,46 @@
+
+        <p>
+          <%= form.label :rule_type %>
+          <%= form.collection_select :rule_type, BadgeService.rules, :first, :last, id: :rule_type, prompt:  t('helpers.select.badge.prompt'), class: 'form-select'  %>
```
как хранилище статичных данных используют константы. Хорошо бы такую константу определить в модели бадж. По смыслу очень подходит

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+      rule_description[rule_id] = rule.description
+    end
+    rule_description
+  end
```
процедурный стиль можно заменить одним методом map
map вернёт изменённый массив 
each_with_object для каких то преоброзований

VidgarVii commented on app/services/finish_any_test_by_first_try_rule_specification.rb:
```diff
@@ -0,0 +1,16 @@
+
+
+  def self.description
+    "За успешное прохождение теста с первого раза"
```
1 контекст всегда хранят в файлах локализации
2 Так же используя файл локализации делает этот метод не нужным. По типу правила можно получать нужный контекст
3 Спецификации ни когда не хранят какие-то данные или описание. Он только решает вопрос удовлетворённости

VidgarVii commented on app/services/finish_any_test_by_first_try_rule_specification.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def satisfies?
+    if  @processed_test.success?
+       ProcessedTest.where("user_id=? and test_id=? and id<>?", @processed_test.user_id, @processed_test.test_id, @processed_test.id).empty?
```
ну почему такой бардак с отступами?(
Не можешь следить сам за этим ни чего страшного, используй рубоком! Ну чтобы был везде порядок!

VidgarVii commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -0,0 +1,25 @@
+  end
+
+  def satisfies?
+    category = Category.where(title:@value).first
```
если ты используешь не `id`, а `title`, то логично будет добавить валидацию уникальности этого поля в модели `Category`
по этому такие проверки лучше делать по "ай-ди" (в большинстве случаев)

VidgarVii commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -0,0 +1,25 @@
+    else
+      tests == user_tests
+    end
+  end
```
разбей это большой сложный не читаемый код на маленькие простые методы с понятными названиями

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def call
+    Badge.all.each do |badge|
```
лучше используй find_each для итераций.
метод .all почти не используют. Так как если в БД окажется большое кол-во данных, то данный метод полностью заполнит оперативную память.  И сервер зависнет

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+    @given_badges
+  end
+  def add_badge(badge)
+    if (@processed_test.user.badges.where("badge_id=?", badge.id).empty? && badge.unique) || !badge.unique
```
везде где возможно читаемей и безопасней использовать хеш синтаксис
`.where(badge_id: badge.id)`

VidgarVii commented on app/services/badge_service.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+  def add_badge(badge)
+    if (@processed_test.user.badges.where("badge_id=?", badge.id).empty? && badge.unique) || !badge.unique
+      @processed_test.user.badges << badge
```
в каждой итерации происходит запрос в базу на вставку баджа.
На много эффективней было бы это делать 1 раз.
Сначала получаешь список заработанных баджей, А затем их добавляешь юзеру

JuicyFlash committed:
Добавлен gem rubocop
Форматирование кода

JuicyFlash committed:
Добавлена свзяь processed_test-given_badges-badges
Изменено место вызова BadgeService (вызывается в момент завершения)
Переработан BadgeService:
  - Выданные баджи добавляются одним запросом к бд
  - find_each вместо each
  - Убран метод контекста правил
Добавлена валидация уникальности поля title в Category
Переработаны спецификации правил:
  - Упрощёны методы satisfies? (разбиты на более простые)
  - find_each вместо each
  - Использование хеш синтаксиса в хелперах where (где смог)
  - Убраны методы контекста правил
Работа с контекстом правил через константу в модели Badge
Контекст правил вынесен в файлы локализации

JuicyFlash committed:
Фикс багов в правилах

BubuntuClu commented on app/services/finish_any_test_by_first_try_rule_specification.rb:
```diff
@@ -0,0 +1,12 @@
+class FinishAnyTestByFirstTryRuleSpecification < AbstractRuleSpecification
+  def satisfies?
+    if @processed_test.success?
+      ProcessedTest.where('user_id=? and test_id=? and id<>?', @processed_test.user_id, @processed_test.test_id,
```
если используешь методы активрекорда, то используй синтаксис который больше подходит актив рекорду `whre(user_id: @processed_test.user_id, ...)`
если надо вставить кусок чистого sql подключай `Arel`

BubuntuClu replied:
кстати, этот запрос вообще будет правильно работать? ты прошел этот тест, в каком случае ты тут получишь `true`?

BubuntuClu commented on app/services/finish_any_test_by_first_try_rule_specification.rb:
```diff
@@ -0,0 +1,12 @@
+
+class FinishAnyTestByFirstTryRuleSpecification < AbstractRuleSpecification
+  def satisfies?
+    if @processed_test.success?
```
`return false unless @processed_test.success?`
чтобы не оборачивать весь код в `if/else` выражение

BubuntuClu commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -0,0 +1,40 @@
+  private
+
+  def category_by_name(category_name)
+    Category.where(title: category_name).first
```
чтобы не исопльзовать `first` используй `find_by`

BubuntuClu commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -0,0 +1,40 @@
+    Test.where(category_id: category_id, ready: true).order('tests.id DESC').pluck(:id)
+  end
+
+  def successful_user_tests_ids_by_category(category_id)
```
почему такой сложный код полуачется? тебе надо сравнить массив тестов категории и массив тестов категории, которые прошел пользователь. что за обработки массивов ты делаешь?
тут же все должно быть два запроса и сравнение между ними результатов

BubuntuClu commented on app/services/finish_tests_by_level_rule_specification.rb:
```diff
@@ -0,0 +1,34 @@
+
+  private
+
+  def successful_user_tests_ids_by_level(level)
```
тут тоже самое что и в категориях. должно быть два запроса и сравнение их результатов. никаких рубишных обработок не должно быть. максимально фильтруй данные на этапе запроса к БД

JuicyFlash committed:
Условия where переписаны на хэш вид
find_by вместо first
Рефакторинг правил - код упрощён

BubuntuClu commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -8,33 +8,23 @@ def satisfies?
     return false if ready_tests_ids.empty?
 
-    ready_tests_ids == successful_user_tests_ids_by_category(category.id)
+    ready_tests_ids == user_processed_tests_by_category(category.id).select(&:success?).pluck(:test_id).uniq
```
`select(&:success?)` это рубишная обработка результатов. чтобы ее не было добавь новое поле в таблицу и выставляй его в `true`когда тест пройден успешно. таким образом ты просто через фильтрацию по полю получишь нужный список, вместо обхода всех тестов и подсчетов результатов у них

JuicyFlash committed:
Добавлено поле successful в таблицу processed_tests
Добавлена логика обработки поля successful в модель ProcessedTest

JuicyFlash committed:
Поле successful добавлено в фильтрацию запросов в правилах

BubuntuClu commented on app/services/finish_tests_by_category_rule_specification.rb:
```diff
@@ -0,0 +1,30 @@
+    category = category_by_name(@value)
+    return false if category.nil?
+
+    ready_tests_ids = ready_tests_ids_by_category(category.id)
```
это еще можно упростить. у тебя в модели тестов есть метод, который делает поиск тестов по названию категории, вот сюда он идеально вписывается

JuicyFlash committed:
Немного упрощено правило по категориям
